### ISSUE TITLE
sshd example with UserSwitch

### ISSUE THREAD

**Issue by pchaigno** [mininet/mininet#783]
### Expected/Desired Behavior:

I'm trying to get the `sshd.py` example to work with a userspace switch. I'm expecting the startup process to run until I get a `mininet>` shell and to be able to connect onto the hosts through SSH. I'm getting the expected result if I use the default Open vSwitch.

### Actual Behavior:

The startup process remains stuck on `*** Waiting for ssh daemons to start` as it cannot join the hosts through SSH. Even with a separate shell, I cannot connect to any of the hosts on port 22, although I can ping all of them successfully.

### Detailed Steps to Reproduce the Behavior:

I changed `sshd.py` to use the userspace switch (the default OF1.3 UserSwitch) and a remote controller (in my case, Floodlight):
```diff
diff --git a/examples/sshd.py b/examples/sshd.py
index 7e0f517..fcf7a16 100755
--- a/examples/sshd.py
+++ b/examples/sshd.py
@@ -21,7 +21,7 @@
 from mininet.net import Mininet
 from mininet.cli import CLI
 from mininet.log import lg, info
-from mininet.node import Node
+from mininet.node import Node, UserSwitch, RemoteController
 from mininet.topolib import TreeTopo
 from mininet.util import waitListening
 
@@ -29,7 +29,7 @@
 def TreeNet( depth=1, fanout=2, **kwargs ):
     "Convenience function for creating tree networks."
     topo = TreeTopo( depth, fanout )
-    return Mininet( topo, **kwargs )
+    return Mininet( topo, controller=lambda a: RemoteController( a, ip="172.30.1.3", port=6653 ), switch=UserSwitch, *kwargs )
 
 def connectToRootNS( network, switch, ip, routes ):
     """Connect hosts to root namespace via switch. Starts network.
```

When I run the script, it blocks on the SSH connections:
```
$ sudo python sshd.py 
*** Creating network
*** Adding controller
*** Adding hosts:
h1 h2 h3 h4 
*** Adding switches:
s1 
*** Adding links:
(s1, h1) (s1, h2) (s1, h3) (s1, h4) 
*** Configuring hosts
h1 h2 h3 h4 
*** Starting controller
c0 
*** Starting 1 switches
s1 
*** Waiting for ssh daemons to start
^CTraceback (most recent call last):
  File "sshd.py", line 84, in <module>
    sshd( net, opts=argvopts )
  File "sshd.py", line 66, in sshd
    waitListening( server=server, port=22, timeout=5 )
  File "build/bdist.linux-x86_64/egg/mininet/util.py", line 606, in waitListening
  File "build/bdist.linux-x86_64/egg/mininet/util.py", line 133, in quietRun
  File "build/bdist.linux-x86_64/egg/mininet/util.py", line 96, in errRun
KeyboardInterrupt
```

### Additional Information:

I checked with `ip a` that all network interfaces have been successfully created and with `top` that the `sshd` processes are waiting.

**Comment by pchaigno** [comment]
**TL;DR** The correct fix here is to use the `TCULink` links when using one of the userspace switches. The correct diff is therefore:
```diff
diff --git a/examples/sshd.py b/examples/sshd.py
index c25f1be..4430fd0 100755
--- a/examples/sshd.py
+++ b/examples/sshd.py
@@ -21,14 +21,15 @@
 from mininet.net import Mininet
 from mininet.cli import CLI
 from mininet.log import lg
-from mininet.node import Node
+from mininet.node import Node, UserSwitch, RemoteController
+from mininet.link import TCULink
 from mininet.topolib import TreeTopo
 from mininet.util import waitListening
 
 def TreeNet( depth=1, fanout=2, **kwargs ):
     "Convenience function for creating tree networks."
     topo = TreeTopo( depth, fanout )
-    return Mininet( topo, **kwargs )
+    return Mininet( topo, controller=lambda a: RemoteController(a, ip="127.0.0.1", port=6653), switch=UserSwitch, link=TCULink, **kwargs )
 
 def connectToRootNS( network, switch, ip, routes ):
     """Connect hosts to root namespace via switch. Starts network.
```
----
By tracing the `kfree_skb` Linux kernel function, I noticed that packets were being dropped because of their **incorrect TCP checksums**. During the TCP handshake for the SSH connection, TCP SYN packets from the root host are delivered to the Mininet hosts and dropped in the Linux kernel TCP stack before reaching the userspace application. The SSH server therefore never answers and the client keeps trying to establish the connection.
I noticed that the packet was properly delivered, and a TCP SYN+ACK packet sent as a response, when I **disabled TX checksum offload** on the root host for the `root-eth0` interface (the virtual interface used to connect from the root host to the Mininet hosts):
```
sudo ethtool -K root-eth0 tx off
```
I needed to do the same on the reverse path for the TCP handshake to complete:
```
pid=xxxx # PID a process in the network namespace of the target Mininet host, e.g., the `sshd -D` process.
sudo ln -s /proc/$pid/ns/net /var/run/netns/$pid # To enable `ip netns exec $pid`.
sudo ip netns exec $pid ethtool -K h1-eth0 tx off # Disables TX checksum offload on the Mininet host interface.
```

The correct fix here is to use the `TCULink`, designed exactly for this purpose, which performs the above steps automatically. My guess here is that, when checksum offload is active, Linux expects something else to compute the checksum---maybe the userspace switch. Since this doesn't happen, the malformed packet is dropped on the receive side.

**Comment by mrVaro** [comment]
I had also the same issue. and for me, i just start ssh service using : 
systemctl start ssh
